### PR TITLE
High quality resampler

### PR DESCRIPTION
**Pull request by kamiyo**
Think of this as rough draft. Taking suggestions to make it robust, error free, and prettier!

For more information:
https://docs.google.com/document/d/1tBEgsJh7QiwNwepXI0eobfK3U8LkJButSyeuFt1degM/edit?usp=sharing

removed: SSE includes

added: 16bit -> float -> 16bit functions
added: linear interpolator and high-quality interpolator functions (including Resampler class)
added: dithering

changed: renamed variables and reformatted a few things to fit with style guide
changed: use s16, u16, s32, u32 etc
changed: store samples and do all computations as floats
changed: volume from 0 - 255

### CODE REVIEW

**Comment by Buddybenj** [comment]
Bump. Any update on this?

**Comment by kamiyo** [comment]
I am (still) all ears for improvements and such!

File: Source/Core/AudioCommon/Mixer.h [hunk 1]
```diff
@@ -107,46 +113,96 @@ class CMixer {
...
 	};
-	MixerFifo m_dma_mixer;
-	MixerFifo m_streaming_mixer;
-	MixerFifo m_wiimote_speaker_mixer;
-	unsigned int m_sampleRate;
+
+	class LinearMixerFifo : public MixerFifo {
+	public:
+		LinearMixerFifo(CMixer* mixer, u32 sample_rate) : MixerFifo(mixer, sample_rate)	{}
+		void Interpolate(u32 left_input_index, float* left_output, float* right_output);
```
Comment (delroth) [inline Source/Core/AudioCommon/Mixer.h]: void Interpolate(...) override;

Same for Windowed*

File: Source/Core/AudioCommon/Mixer.h [hunk 2]
```diff
@@ -107,46 +113,96 @@ class CMixer {
...
+		float        m_fraction;
 	};
-	MixerFifo m_dma_mixer;
-	MixerFifo m_streaming_mixer;
-	MixerFifo m_wiimote_speaker_mixer;
-	unsigned int m_sampleRate;
+
+	class LinearMixerFifo : public MixerFifo {
+	public:
+		LinearMixerFifo(CMixer* mixer, u32 sample_rate) : MixerFifo(mixer, sample_rate)	{}
```
Comment (delroth) [inline Source/Core/AudioCommon/Mixer.h]: Why explicitly define the default constructor?

File: Source/Core/AudioCommon/Mixer.h [hunk 3]
```diff
@@ -107,46 +113,96 @@ class CMixer {
...
+
+		float        m_num_left_i;
+		float        m_fraction;
 	};
-	MixerFifo m_dma_mixer;
-	MixerFifo m_streaming_mixer;
-	MixerFifo m_wiimote_speaker_mixer;
-	unsigned int m_sampleRate;
+
+	class LinearMixerFifo : public MixerFifo {
```
Comment (delroth) [inline Source/Core/AudioCommon/Mixer.h]: { on next line.

**Comment by delroth** [comment]
Fix these few remaining style issues and I think this can be merged. Thanks!

**Comment by delroth** [comment]
Oh, and clean your commit history. It's really dirty right now. Squash merge into the commit introducing the change, and reword it so that it explains what the change does better.

**Comment by oddMLan** [comment]
@kamiyo 
Bump :+1:

**Comment by Buddybenj** [comment]
Do you mind rewording the commit message to be descriptive and rebasing this PR?

**Comment by kamiyo** [comment]
What do you mean by descriptive? I'll try rebasing again.

**Comment by Buddybenj** [comment]
The commit just says "For more information:" how about "High Quality Resampler"

**Comment by kamiyo** [comment]
haha I guess the comments I wrote under it didn't show up? I'll try to fix that and add even more =)

File: Source/Core/AudioCommon/Mixer.cpp [hunk 4]
```diff
@@ -51,134 +105,146 @@ unsigned int CMixer::MixerFifo::Mix(short* samples, unsigned int numSamples, boo
...
+	float ratio = aid_sample_rate / (float) m_mixer->m_sample_rate;
+
+	float l_volume = (float) m_lvolume / 255.f;
+	float r_volume = (float) m_rvolume / 255.f;
+
+	// for each output sample pair (left and right),
+	//   linear interpolate between current and next sample
+	//   increment output sample position
+	//   increment input sample position by ratio, store fraction
+	// QUESTION: do we need to check for NUM_CROSSINGS samples before we interpolate?
```
Comment (kamiyo) [inline Source/Core/AudioCommon/Mixer.cpp]: Can someone look at this and let me know if I need to check write_index - read_index > NUM_CROSSINGS

**Comment by ADormant** [comment]
What is the status on this ?

**Comment by kamiyo** [comment]
Mostly done on my end, I think.

**Comment by delroth** [comment]
It needs a rebase before it can be merged.

File: Source/Core/AudioCommon/Mixer.cpp (hunk 4 above)
Comment (delroth) [inline Source/Core/AudioCommon/Mixer.cpp]: @degasus

**Comment by ADormant** [comment]
Seems to be already rebased ?

**Comment by ADormant** [comment]
Can it be merged now ?

File: Source/Core/AudioCommon/Mixer.cpp [hunk 5]
```diff
@@ -51,134 +105,146 @@ unsigned int CMixer::MixerFifo::Mix(short* samples, unsigned int numSamples, boo
...
+	m_streaming_mixer.Mix(m_output_buffer, num_samples, consider_framelimit);
+	m_wiimote_speaker_mixer.Mix(m_output_buffer, num_samples, consider_framelimit);
+
+	// dither and clamp
+	for (u32 i = 0; i < num_samples * 2; i += 2)
+	{
+		float l_output = m_output_buffer[i + 1];
+		float r_output = m_output_buffer[i];
+		TriangleDither(&m_output_buffer[i + 1], &m_output_buffer[i]);
+
```
Comment (Tinob) [inline Source/Core/AudioCommon/Mixer.cpp]: You are not really applying dither here as  the output for dithering is the buffer and you have already readed the values from the variables. also dithering should be aplied to the scaled variables so first you have to multiply the sanple  by 32768 then apply dithereng, clamp  and finally cast to s16.
you can see that applied here:
http://audacity.googlecode.com/svn/audacity-src/trunk/src/Dither.cpp

Comment (kamiyo) [inline Source/Core/AudioCommon/Mixer.cpp]: Whoops, dunno what happened here. I reread the linked src code and I'll fix this. Thanks.

**Pull request merged**